### PR TITLE
sonic-mgmt / IXIA : T2 test  to verify ECN Counter operation pre and post port state toggle

### PR DESCRIPTION
**Pull request by sreejithsreekumaran**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [x] 202405

#### What is the motivation for this PR?
MIGSMSFT-630 Add sonic-mgmt / IXIA test for cisco platforms to verify ECN Counter operation.

#### How did you do it?

Create two flows one for each lossless TC per Ixia port each at half of 99.98% rate

- Inject traffic and read ECN Counters and verify it incremented.

- flap the interface state (shut, wait for 0.2 seconds, no shut the interfaces of Ixia)

- restart the traffic streams

- verify that the ECN counters increment from the baseline value (new values read MUST NOT be zero)

#### How did you verify/test it?

Verified on T2 DUT setup

```
============================================================================================================================================ PASSES =============================================================================================================================================
____________________________________________________________________________________________________________ test_ecn_marking_port_toggle[multidut_port_keys=single-dut-multi-asic] _____________________________________________________________________________________________________________
____________________________________________________________________________________________________________ test_ecn_marking_port_toggle[multidut_port_keys=single-dut-single-asic] ____________________________________________________________________________________________________________
---------------------------------------------------------------------------------------------------- generated xml file: /run_logs/ixia/17467/2024-11-15-16-20-28/tr_2024-11-15-16-20-28.xml ----------------------------------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------------------------ live log sessionfinish -------------------------------------------------------------------------------------------------------------------------------------
16:34:46 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
==================================================================================================================================== short test summary info ====================================================================================================================================
PASSED snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py::test_ecn_marking_port_toggle[multidut_port_keys=single-dut-multi-asic]
PASSED snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py::test_ecn_marking_port_toggle[multidut_port_keys=single-dut-single-asic]
=========================================================================================================================== 2 passed, 5 warnings in 856.18s (0:14:16) ==========================================================================================================================
```

### CODE REVIEW

**Comment by sreejithsreekumaran** [comment]
@rraghav-cisco  @abdosi @sdszhang

**Review by alpeshspatel — APPROVED** [review]

**Comment by sdszhang** [comment]
@abdosi can you help to review this one? Thanks.

File: tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        # Collect port names from rx_ports and tx_ports into a set for uniqueness
+        all_ports_set = set(port['port_name'] for port in rdma_ports['rx_ports'] + rdma_ports['tx_ports'])
+        pytest_assert(len(all_ports_set) >= tx_port_count + rx_port_count,
+                      'MULTIDUT_PORT_INFO doesn\'t have at least {} unique ports for \
+                      testbed {}, subtype {} in variables.py'.
+                      format(tx_port_count + rx_port_count, MULTIDUT_TESTBED, testbed_subtype))
+
+        logger.info('Running test for testbed subtype: {}'.format(testbed_subtype))
+
+        if is_snappi_multidut(duthosts):
```
Comment (sdszhang) [inline tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py:83]: Raghav has a recent commit which moved these port selection into fixture for easy setup.  https://github.com/sonic-net/sonic-mgmt/pull/15099/

Do you want to adopt the change in this PR or address it in another PR?

File: tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        get_snappi_ports (pytest fixture): gets snappi ports and connected DUT port info and returns as a list
+    Returns:
+        N/A
+    """
+
+    for testbed_subtype, rdma_ports in multidut_port_info.items():
+        tx_port_count = 2
+        rx_port_count = 1
+        snappi_port_list = get_snappi_ports
+        pytest_assert(MULTIDUT_TESTBED == tbinfo['conf-name'],
```
Comment (sdszhang) [inline tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py:58]: This line needs to be removed.  as T0/T1 doesn't need to check this.

File: tests/snappi_tests/multidut/ecn/files/multidut_helper.py [hunk 3]
```diff
@@ -181,3 +232,192 @@ def run_ecn_test(api,
...
+    post_ctr_4 = get_npu_voq_queue_counters(duthost, dut_port, test_prio_list[1])
+
+    ecn_counters = [
+        (init_ctr_3, post_ctr_3),
+        (init_ctr_4, post_ctr_4)
+    ]
+
+    verify_ecn_counters(ecn_counters)
+
+    # Get the current configuration
```
Comment (sdszhang) [inline tests/snappi_tests/multidut/ecn/files/multidut_helper.py:383]: better to move this part of the code into a function?

File: tests/snappi_tests/multidut/ecn/files/multidut_helper.py [hunk 4]
```diff
@@ -21,6 +23,55 @@
...
+    pytest_assert(flow4_total > 0,
+                  'Queue 4 counters at start {} at end {} did not increment{}'.format(
+                   init_ctr_4['SAI_QUEUE_STAT_PACKETS'], post_ctr_4['SAI_QUEUE_STAT_PACKETS'], toggle_msg))
+
+    flow3_ecn = post_ctr_3['SAI_QUEUE_STAT_WRED_ECN_MARKED_PACKETS'] -\
+        init_ctr_3['SAI_QUEUE_STAT_WRED_ECN_MARKED_PACKETS']
+    flow4_ecn = post_ctr_4['SAI_QUEUE_STAT_WRED_ECN_MARKED_PACKETS'] -\
+        init_ctr_4['SAI_QUEUE_STAT_WRED_ECN_MARKED_PACKETS']
+
+    pytest_assert(flow3_ecn > 0,
```
[the remark this answers is not in the document]
Comment (sreejithsreekumaran) [reply tests/snappi_tests/multidut/ecn/files/multidut_helper.py:66]: In this specific port toggle test case, we are only making sure the ECN marking is happening. 

We will have separate test to check the percentage of marking probably based on DWRR.

File: tests/snappi_tests/multidut/ecn/files/multidut_helper.py (hunk 3 above)
Comment (sreejithsreekumaran) [reply tests/snappi_tests/multidut/ecn/files/multidut_helper.py:383]: done

File: tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py (hunk 1 above)
Comment (sreejithsreekumaran) [reply tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py:83]: integrated the change

File: tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py (hunk 2 above)
Comment (sreejithsreekumaran) [reply tests/snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py:58]: This is gone due to PR 15099 changes

**Comment by abdosi** [comment]
@sreejithsreekumaran : please resolve conflict.

**Pull request merged**